### ISSUE TITLE
sync sessions between two instances

### ISSUE THREAD

**Issue by SchSeba** [yyyar/gobetween#164]
I will like to know if the project support or will support the option to create a cluster with a number of gobetween instances.(cluster mode)

**Comment by nickdoikov** [comment]
hello @SchSeba .
What you mean under "cluster mode"?
you can place gb config somewhere , and point all gb instances to this config . For example you can use consul or http for that .
or you mean dynamic servers sync?

**Comment by SchSeba** [comment]
Hi @nickdoikov I mean to sync tcp sessions lets said I put a keepalived to have the same ip between 2 machines so when one goes down the ip address will jump to the secondary machine but I will like to also continue the same tcp sessions.

**Comment by SchSeba** [comment]
@yyyar this is a feature that will be possible?

I am using a keepalived to have a floating ip address but if one machine is go down all the active connections are close and the client needs to reconnect

**Comment by SchSeba** [comment]
@yyyar Yes haproxy have the ability to sync sessions between members using a peer configuration.

http://www.haproxy.org/download/1.5/doc/configuration.txt Section 3.5 Peers

**Comment by SchSeba** [comment]
@yyyar I was trying to looking into this and I am not able to create a net.Conn object.

There needs to be a way to create a file descriptor manually but I cant find one and then I can call net.FileConn() to get the net.Conn object.

You have any idea about it?

**Comment by ryan8613** [comment]
Let "node" equal a gobetween instance.

To me, nodes would communicate TCP states to neighboring nodes. Connection states would time out of course. (it would be OK that only TCP or TLS were supported for HA - UDP is mostly just a passthru anyways and handled by the VIP mechanism)

The network stack (sharing of IP[s]) would be handled by the appropriate OS feature (Linux - MASQ, Windows - NLB, etc.)

When or if a node either: 
    1. Recognized its listening VIP is no longer active on the host. 
    2. Recognized ALL upstream servers are no longer active.
    3. Ran in to a health issue or gobetween exception.
The node would funnel requests/responses to remaining active nodes, until the point it self-terminates due to being "inactive" (which basically just means, all of its known connection states have timed out). Beyond the funnel, nodes would either, respectfully:
    1. Wait for its VIP to become active again.
    2. Continue monitoring for upstream servers to see when they again become active.
    3. Exit as gracefully as possible at this point, or clear memory and restart, depending on the situation.

The protocol between nodes could be a million different approaches, but usually the KISS approach wins out for success. Something small, efficient, and with encryption capability for the security nuts.

**Comment by siacali** [comment]
I've got simple HA failover running in production with gobetween using ucarp on Ubuntu.  There's no session sync, and so in-progress connections are dropped, but the secondary kicks in fast enough that users do not generally perceive it.
